### ISSUE TITLE
ESI decreases performance 

### ISSUE THREAD

**Issue by Norfavrell** [nexcess/magento-turpentine#902]
Hey,

I have installed the latests Turpentine, made a small modification to ESI block URLs (https issue #684) and fixed some bad layout structure from my theme. I have been running stress tests over the few days and it seems that ESI does not actually offer any decrease in load times and does in fact increase them. 

## Setup
_ESI connections not shown. Magento unsecure URL is https so client doesn't normally connect to nginx:80_
![Setup](https://i.imgur.com/4Ga3Kbt.jpg)

### Without ESI
![Without ESI](https://i.imgur.com/2nKm3Vn.png) 
### With ESI
_Don't quite understand how they are higher since all pages should be cached by FPC mechanism anyway yet I am consistently seeing similar results_
![With ESI](https://i.imgur.com/MnUnNmh.png)

_The question is_: do I really need ESI enabled with Turpentine? Since it doesn't seem to provide any real performance benefit and requires a lot of bug fixing (with the theme/modules I use) I would rather have it disabled. I don't see, conceptually, why disabling this would cause any problems yet you advise against it on 'Cache Managment' page. 

_NB: I'm running tests using [bombard](https://github.com/allardhoeve/bombard) (siege wrapper), all after warming up the cache. The server is (mostly) idle at the time of testing - although slight variations are seen in the results. I believe my test methodology to be slightly flawed at places but I do think they are useful for the sake of comparison._

**Comment by aricwatson** [comment]
I suppose you don't really need ESI enabled - however you do lose some options if it's turned off. One of the key things you'll lose if you disable ESI is the option to have some blocks cached privately vs publicly, setting specific TTLs for different blocks, using ajax loading, etc.

I would test this change very carefully, especially watching for problems with non-public information being cached. Let us know what you find!

**Comment by miguelbalparda** [comment]
@Norfavrell I checked this and I actually think nothing will be cached since you are not storing sessions in any way with your tests. We have several fixes and the one active by default is a pass, which means the first request will always miss or pass Varnish. This is because we need to generate a session and bypassing Varnish is one way to achieve this. This can be changed as stated in #846 but the default behaviour is the miss for the first request. What do you think?

**Comment by BarryCarlyon** [comment]
I'm with @miguelbalparda here @Norfavrell your tests appear to indicate that each siege from bombard is a "new user" and thus every time getting a cold page/section of the page, it's not a "fair" analysis of a overall Magento/Varnish install/user interaction with the website.

As page one has to generate the session/form key, which is preserved for each page load for the session which then hits the cache "nicely" where as here there is no session data preservation between runs. ESI includes tends to be where the form key comes in (for basket addition etc) but Turp's works in either mode.

So yes the test is flawed.

Generally speaking ESI is more useful on, as you can cache separate parts of a page independently of each other as apposed to Full Page Caching. Sure it comes with a minor speed loss, depending on how many ESI blocks you have per page.

**Comment by miguelbalparda** [comment]
Any news on this @Norfavrell?

**Comment by Norfavrell** [comment]
Sorry @miguelbalparda, didn't have a chance to progress any further with this as of yet. As for now I have enabled ESI and fixed all the broken blocks (wasn't that difficult afterall). 

I have scheduled a few tasks to correct the methodology of our performance tests and finish the configuration for the next sprint - I will keep you updated when I have any news.

Again, sorry about the delay. Everything Turpentine does for us is currently functional and since we are approaching the initial release we had to focus the development effort on functionality-breaking issues.
